### PR TITLE
feat(rest-connector): Add support for proxy auth

### PR DESCRIPTION
**Pull request by ztefanie**
## Description
Added support for setting proxy via env var and to add proxy auth.

Proxy config can be set via system properties or via env vars. Setting via env vars is currently handled by using an RFC3986 string, e.g. http://my-user:demo@localhost:3128/
This means it is currently not possible to set nonProxyHosts via env vars and username and passwords must be url safe.
I think we could also change this, with some effort, to setting multiple env vars. I decided to go for this version, assuming setting via env vars is more for testing / fast setups, but not complex production setups, as setting passwords in ENV vars is worse than setting in system properties. If you disagree please let me know, then i will adjust it.

I tested different use cases manually, with the following results:
image

I tried to add Unit tests for nonProxyHosts but couldn't get it working as mocking only one host / localhost is possible with Wiremock. If you think nonProxyHost unit test should be added, please give me a hint how to achieve this.

Not tested for https.

I will add this to the docs, once this is approved or at least we agreed on how to set it with env vars.

## Related issues

closes https://github.com/camunda/team-connectors/issues/989

## Checklist

- [x] PR has a **milestone** or the `no milestone` label.

### CODE REVIEW

File: connectors/http/http-base/src/test/resources/squid.conf [hunk 1]
```diff
@@ -6,6 +6,19 @@ acl Safe_ports port 80
...
 acl CONNECT method CONNECT
 
-# Allow all outgoing HTTP/HTTPS traffic
-http_access allow all
+# Define the authentication method
+auth_param basic program /usr/lib/squid/basic_ncsa_auth /etc/squid/passwords
+auth_param basic realm Squid Proxy
 
+# Allow unauthenticated users on /path
+acl no_auth_url url_regex -i path
```
Comment (ztefanie) [inline connectors/http/http-base/src/test/resources/squid.conf:14]: so with this setting every path is needs username and password except "/path". I would rather have only "/protected" be protected and everything else allowed, but I am not sure if this is possible. So if you have an idea let me know.

**Comment by sbuettner** [comment]
@ztefanie We currently only document environment variables for our component configuration and we should therefore continue to support using them also for this use-case: https://docs.camunda.io/docs/next/self-managed/connectors-deployment/connectors-configuration/

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                Integer.parseInt(System.getProperty(protocol + ".proxyPort"))),
+            new UsernamePasswordCredentials(
+                System.getProperty(protocol + ".proxyUser"),
+                System.getProperty(protocol + ".proxyPassword").toCharArray()));
+      } else if (systemEnvIsSetForProtocol(protocol)) {
+        CustomProxy proxyData = getProxySettingsFromEnvVar(protocol);
+        provider.setCredentials(
+            new AuthScope(proxyData.host, proxyData.port),
+            new UsernamePasswordCredentials(proxyData.username, proxyData.password.toCharArray()));
+      }
```
Comment (sbuettner) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:38]: Would it be possible to have a dedicated `ProxyConfig` data class and a single place (a factory) that is responsible for creating it? 

It would be easier to follow and having a single place where data is read either from System.properties and EnvVars would make it easier to debug.

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+  private static boolean doesTargetMatchNonProxy(String protocol, String requestUri) {
+    String nonProxyHostsEnvVar =
+        System.getenv("CONNECTOR_" + protocol.toUpperCase() + "_PROXY_NON_PROXY_HOSTS");
+    return (nonProxyHostsEnvVar != null
+        && Arrays.stream(nonProxyHostsEnvVar.split("\\|"))
+            .anyMatch(
+                nonProxyHost ->
+                    requestUri.matches(nonProxyHost.replace(".", "\\.").replace("*", ".*"))));
+  }
```
Comment (sbuettner) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:74]: This check should be done against a proxy config without requiring to fetch the data from the environment. Possibly using a dedicated `ProxyConfig` as mentioned above.

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java (hunk 2 above)
Comment (sbuettner) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:38]: I just saw that we have `CustomProxy` available. Why not use it here as well:
```
        provider.setCredentials(
            new AuthScope(
                System.getProperty(protocol + ".proxyHost"),
                Integer.parseInt(System.getProperty(protocol + ".proxyPort"))),
            new UsernamePasswordCredentials(
                System.getProperty(protocol + ".proxyUser"),
                System.getProperty(protocol + ".proxyPassword").toCharArray()));
```

Comment (ztefanie) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:38]: You are right, that would make it easier to read. I will adjust it.

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/CustomApacheHttpClient.java [hunk 4]
```diff
@@ -97,14 +99,22 @@ private static PoolingHttpClientConnectionManager createConnectionManager() {
   @Override
   public HttpCommonResult execute(
       HttpCommonRequest request, @Nullable ExecutionEnvironment executionEnvironment) {
-    var apacheRequest = ApacheRequestFactory.get().createHttpRequest(request);
     try {
+      var apacheRequest = ApacheRequestFactory.get().createHttpRequest(request);
+      ProxyHandler proxyHandler =
+          new ProxyHandler(apacheRequest.getScheme(), apacheRequest.getUri().getHost());
```
Comment (sbuettner) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/CustomApacheHttpClient.java:105]: Looks like we are creating a new proxy handler for every new request. I think this should be done once.

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+  public CredentialsProvider getCredentialsProvider(String uncheckedProtocol) {
+    BasicCredentialsProvider provider = new BasicCredentialsProvider();
+    if (StringUtils.isNotBlank(host)
+        && StringUtils.isNotBlank(user)
+        && ArrayUtils.isNotEmpty(password)) {
+      provider.setCredentials(
+          new AuthScope(host, port), new UsernamePasswordCredentials(user, password));
+    }
+    return provider;
```
Comment (sbuettner) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:76]: What happens if the password or another entry is empty? Should we log a warning or something?

Otherwise this will silently lead to a credentials provider without credentials.

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  public HttpRoutePlanner getRoutePlanner(String uncheckedProtocol, HttpHost proxyHost) {
+    if (sourceIsSystemProperties) {
+      return new SystemDefaultRoutePlanner(
+          DefaultSchemePortResolver.INSTANCE, ProxySelector.getDefault());
+    } else if (proxyHost != null) {
+      return new DefaultProxyRoutePlanner(proxyHost);
+    }
+    return null;
```
Comment (sbuettner) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:100]: Would be great to add some debug logs here. Otherwise it might be intransparent which config was applied.

Comment (ztefanie) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:100]: Added debug logging.

File: connectors/http/http-base/src/test/resources/squid.conf (hunk 1 above)
Comment (johnBgood) [reply connectors/http/http-base/src/test/resources/squid.conf:14]:
```
acl protected_path url_regex -i ^/protected

auth_param basic program /usr/lib/squid/basic_ncsa_auth /etc/squid/passwords
acl authenticated_user proxy_auth REQUIRED

http_access deny protected_path !authenticated

http_access allow all
```

Not sure but you might try this approach

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  record ProxyDetails(
+      String protocol,
+      String host,
+      int port,
+      String user,
+      char[] password,
+      String nonProxyHosts,
+      boolean sourceIsSystemProperties) {}
+
+  private Map<String, ProxyDetails> map = new HashMap<>();
```
Comment (johnBgood) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:33]: `private final Map<String, ProxyDetails> map = new HashMap<>();`

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(ProxyHandler.class);
+
+  public ProxyHandler() {
+    try {
+      setFromEnvVars("http");
```
Comment (johnBgood) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:38]: I think I would create a global variable `PROTOCOLS = http, https`

and here something like:
```java
Map<String, ProxyDetails> config = getConfigFromSystemProperties().orElse( () -> getConfigFromEnvVars());

// Set the 2 different CredentialsProvider as class members to avoid creating them on every request
```

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+              .orElseGet(() -> new char[0]);
+      String nonProxyHosts = System.getProperty(protocol + ".nonProxyHosts");
+      map.put(
+          protocol, new ProxyDetails(protocol, host, port, user, password, nonProxyHosts, true));
+      LOGGER.debug("Proxy settings applied from system properties for {}", protocol);
+    }
+  }
+
+  public CredentialsProvider getCredentialsProvider(String protocol) {
+    BasicCredentialsProvider provider = new BasicCredentialsProvider();
```
Comment (johnBgood) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:83]: as I said here I would not create a `CredentialsProvider` I would just return the store value for this protocol.

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+      provider.setCredentials(
+          new AuthScope(p.host(), p.port()),
+          new UsernamePasswordCredentials(p.user(), p.password()));
+    }
+    return provider;
+  }
+
+  public HttpHost getProxyHost(String protocol, String requestUrl) {
+    ProxyDetails p = map.get(protocol);
+    return p == null || doesTargetMatchNonProxy(protocol, requestUrl)
```
Comment (johnBgood) [inline connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:98]: I'd add some logs here showing what is actually used

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java (hunk 9 above)
Comment (ztefanie) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:83]: And the CredentialsProvider should then be created in the execute function of the CustomApacheHttpClient?

Comment (johnBgood) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:83]: I would create them in the constructor, you can see my comment above (which is not super clear maybe sorry about that :D )

File: connectors/http/http-base/src/test/resources/squid.conf (hunk 1 above)
Comment (ztefanie) [reply connectors/http/http-base/src/test/resources/squid.conf:14]: Thanks for your suggestion! That was actually my first idea as well, but in my tests, it enforced authentication for all paths. I tweaked it a bit, and now it’s working as expected. Really appreciate your input!

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java (hunk 9 above)
Comment (ztefanie) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:83]: Ah of course, sorry have missed that. Thanks for clarifying. I adjusted it.

File: connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+import org.apache.hc.client5.http.impl.DefaultSchemePortResolver;
+import org.apache.hc.client5.http.impl.auth.BasicCredentialsProvider;
+import org.apache.hc.client5.http.impl.routing.DefaultProxyRoutePlanner;
+import org.apache.hc.client5.http.impl.routing.SystemDefaultRoutePlanner;
+import org.apache.hc.client5.http.routing.HttpRoutePlanner;
+import org.apache.hc.core5.http.HttpHost;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProxyHandler {
```
[the remark this answers is not in the document]
Comment (ztefanie) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:23]: Move it to a separate file

[the remark this answers is not in the document]
Comment (ztefanie) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:25]: Really good point :fire:  I added more tests especially for nonProxyHost matching check and found some wrongly handed edge case.

[the remark this answers is not in the document]
Comment (johnBgood) [reply connectors/http/http-base/src/main/java/io/camunda/connector/http/base/client/apache/ProxyHandler.java:25]: Love it, nice!

**Review by johnBgood — APPROVED** [review]
💯

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Test required in (you can check both):
  - [ ] SaaS
  - [ ] SM
- [ ] Create a new test for this feature in our [Pre Release Tests](https://modeler.ultrawombat.com/projects/017779a8-9459-4feb-8b47-eeef8b2717ef--connectors-pre-release-test) for @Szik to be able to test

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*
